keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java

```diff
@@ -264,4 +236,28 @@ public PartialDerivatives clone() {
         }
         return clone;
     }
+
+    private static DoubleTensor increaseRankByAppendingOnesToShape(DoubleTensor lowRankTensor, int desiredRank) {
+        return increaseRankByPaddingOnes(lowRankTensor, desiredRank, true);
+    }
+
+    private static DoubleTensor increaseRankByPrependingOnesToShape(DoubleTensor lowRankTensor, int desiredRank) {
```

Comment:
Should we move these to `TensorShape` as public static helpers?

Then we can change:

```
        if (!this.isObserved()) {
            dLogPdInputs.putWithRespectTo(getId(), dLogPdx.reshape(concat(SCALAR_SHAPE, dLogPdx.getShape())));
        }
```

to:

```
        if (!this.isObserved()) {
            dLogPdInputs.putWithRespectTo(getId(), increaseRankByPrependingOnesToShape(dLogPdx, dLogPdx.getRank() + 2));
        }
```